### PR TITLE
Attempt to deal gracefully with Chunk Overflow problem in BnP.

### PR DESCRIPTION
**Pull request by FelixGV**

### CODE REVIEW

File: contrib/hadoop-store-builder/src/java/voldemort/store/readonly/mr/HadoopStoreBuilder.java [hunk 1]
```diff
@@ -287,14 +235,41 @@ public void build() {
...
             logger.info("Building store...");
-            RunningJob job = JobClient.runJob(conf);
 
-            // Once the job has completed log the counter
-            Counters counters = job.getCounters();
+            // The snipped below copied and adapted from: JobClient.runJob(conf);
+            // We have more control in the error handling this way.
+
+            JobClient jc = new JobClient(conf);
+            RunningJob runningJob = jc.submitJob(conf);
```
Comment (arunthirupathi) [inline contrib/hadoop-store-builder/src/java/voldemort/store/readonly/mr/HadoopStoreBuilder.java]: can we name the variable firstJob ?

File: contrib/hadoop-store-builder/src/java/voldemort/store/readonly/mr/HadoopStoreBuilder.java [hunk 2]
```diff
@@ -287,14 +235,41 @@ public void build() {
...
+                    long mapOutputBytes = counters.getCounter(Task.Counter.MAP_OUTPUT_BYTES);
+                    int numChunks = conf.getInt(AbstractStoreBuilderConfigurable.NUM_CHUNKS, 1);
+                    long averageNumberOfBytesPerChunk = mapOutputBytes / numChunks / cluster.getNumberOfPartitions();
+                    if (averageNumberOfBytesPerChunk > HadoopStoreWriter.MAX_CHUNK_SIZE) {
+                        logger.warn("The number of bytes per chunk is too high. averageNumberOfBytesPerChunk = " +
+                                averageNumberOfBytesPerChunk + ". Consider using a smaller chunk size. " +
+                                "The MapReduce job will be retried one more time with adjusted settings.");
+                        setNumChunksAndNumReduceTasks(mapOutputBytes, conf);
+                        try {
+                            RunningJob newJob = JobClient.runJob(conf);
```
Comment (arunthirupathi) [inline contrib/hadoop-store-builder/src/java/voldemort/store/readonly/mr/HadoopStoreBuilder.java]: Can we name this variable retryJob ?

File: contrib/hadoop-store-builder/src/java/voldemort/store/readonly/mr/HadoopStoreBuilder.java (hunk 1 above)
Comment (arunthirupathi) [inline contrib/hadoop-store-builder/src/java/voldemort/store/readonly/mr/HadoopStoreBuilder.java]: When the job failed, it did not proceeded into the next phase in the old code ? Does it throw an exception ? Have you validated this new code ?

File: contrib/hadoop-store-builder/src/java/voldemort/store/readonly/mr/HadoopStoreBuilder.java [hunk 3]
```diff
@@ -287,14 +235,41 @@ public void build() {
...
+                        } catch (IOException e) {
+                            throw new VoldemortException("Even with adjusted settings, BnP's MapReduce job failed.", e);
+                        }
+                    } else {
+                        logger.error("Job Failed: " + runningJob.getFailureInfo());
+                        throw new VoldemortException("BnP's MapReduce job failed.");
+                    }
+                }
+            } catch (InterruptedException ie) {
+                Thread.currentThread().interrupt();
```
Comment (arunthirupathi) [inline contrib/hadoop-store-builder/src/java/voldemort/store/readonly/mr/HadoopStoreBuilder.java]: Why is an interruptedException caught here, if throwing the interruptedException from here fails the code, does it need to be handled ?

**Comment by arunthirupathi** [comment]
Left some comments, but can you add what is the problem and what is the new behavior for this change.

**Comment by FelixGV** [comment]
This is superseded by #439.

**Pull request closed without merging**